main.go

```diff
@@ -201,49 +206,44 @@ func getActiveMiners(db *pg.DB, fp *filecoinPeer) (<-chan minerInfo, error) {
 		return nil, err
 	}
 
-	found := make(chan minerInfo)
-	var wg sync.WaitGroup
-	minersFound := 0
+	var (
+		found      = make([]minerInfo, len(activeMiners))
+		peerIDs    = make(chan minerWithDecodedPeerID, 50)
+		wg         = sync.WaitGroup{}
+		maxWorkers = runtime.GOMAXPROCS(0)
+		sem        = semaphore.NewWeighted(int64(maxWorkers))
+	)
+	for i := range activeMiners {
 
-	for i, activeMiner := range activeMiners {
-		i, activeMiner := i, activeMiner
-		wg.Add(1)
+		miner := activeMiners[i]
+
+		decodedPid, err := peer.Decode(miner.PeerID)
+		if err != nil {
+			log.Warnw("could not decode peer ID ", "ID", miner.PeerID, "error", err)
+		}
+		peerIDs <- minerWithDecodedPeerID{miner: miner, decodedPeerID: decodedPid}
 
+		wg.Add(1)
 		go func() {
 			defer wg.Done()
-
-			ctx, cancel := context.WithTimeout(context.Background(), time.Duration(15*time.Second))
-			defer cancel()
-
-			decodedPid, err := peer.Decode(activeMiner.PeerID)
-			if err != nil {
-				log.Warnw("could not decode peer ID ", "ID", activeMiner.PeerID, "error", err)
-				return
+			if err := sem.Acquire(fp.ctx, 1); err != nil {
+				log.Warnw("Failed to acquire semaphore: ", err)
 			}
-
-			info, err := fp.dht.FindPeer(ctx, decodedPid)
+			defer sem.Release(1)
+			f, err := fp.findPeersWithDHT(<-peerIDs)
 			if err != nil {
-				log.Warnw("could not find peer for ", "ID", activeMiner.PeerID, "error", err)
-			} else {
-				minersFound++
-				peerAddrs := info.Addrs
-				peerAddrsAsStr := make([]string, len(peerAddrs))
-				for i, p := range peerAddrs {
-					peerAddrsAsStr[i] = p.String()
-				}
-				activeMiners[i].MultiAddresses = peerAddrsAsStr
-				activeMiners[i].Source = "dht"
-				found <- activeMiners[i]
+				return
 			}
+			found = append(found, f)
 		}()
 	}
 
 	go func() {
 		wg.Wait()
```

Comment:
Waiting in a goroutine makes the waiting useless as nothing blocks or is meant to happen after waiting.